Motion Planning, Gym Wrapper, Compose Agents, Hab2 Bug fixes

ASzot opened the pull request:
## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- Please link to an existing issue here if one exists. -->
<!--- (we recommend to have an existing issue for each pull request) -->

The next version of the Habitat 2.0 release, this PR adds motion planning, an OpenAI Gym wrapper, bug fixes, and the initial set of features necessary to compose agent policies to record demonstrations with motion planning. 

## How Has This Been Tested

<!--- Please describe here how your modifications have been tested. -->
- Added new tests to `test/test_rearrange_task.py` 
- Added a non-deterministic test to `test/test_motion_plan.py`

## Types of changes

Habitat Baselines Changes:
- Support rendering images from observation with unequal dimensions. 
- Added benchmark_render.py, a new version of benchmark.py which supports rendering videos and saving trajectories to a file. 
- Added Gym wrapper with config for which parts of the observation to get. 
- Added SPA pipeline from Habitat 2.0 paper. 
- Added point cloud code for Planning in PyBullet. 

HAB Task Changes:
- Added more detail to the collision check utilities return values. 
- Added actions for setting absolute kinematic state and target joint states. 
- Updated interactive play script with text. 
- Added separate IK config for the interactive play script. 
- Refactored measurements code to break down the reward signal and reuse measurements for the reward and success signal. 
- Refactored robo versus robot naming. 

Bug fixes:
- Fix bug with size of the joint observation space. 
- Base transform was not being properly calculated. 

<!--- What types of changes does your code introduce? Leave all the items that apply: -->
- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

ASzot commented:
Example of motion planning output: 

Uploading episode=None-ckpt=0-rearrangepick_success=1.00.mp4…

Skylion007 commented on habitat/tasks/rearrange/actions.py:
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
+        self._sim.robot.arm_motor_pos = set_pos
+        if should_step:
+            return self._sim.step(HabitatSimActions.ARM_ABS_POS)
+        else:
```
Else after return is discouraged. Flake8 should complain.

mathfac replied:
Do you mean it should be `return None` without else?

ASzot commented:
Example of motion planning output: 



https://user-images.githubusercontent.com/10453381/125521302-784cc650-9710-46ec-a29e-6b6a2474d8db.mp4

Skylion007 commented on habitat_baselines/agents/benchmark_render.py:
```diff
@@ -0,0 +1,153 @@
+
+        if self._traj_save_path is not None:
+            save_dir = osp.dirname(self._traj_save_path)
+            if not osp.exists(save_dir):
+                os.makedirs(save_dir)
```
```suggestion
            os.makedirs(save_dir, exist_ok=True)
```

Skylion007 commented on habitat_baselines/motion_planning/grasp_generator.py:
```diff
@@ -0,0 +1,341 @@
+
+class GraspGenerator:
+    def __init__(
+        self,
```
missing types

Skylion007 commented on habitat_baselines/motion_planning/mp_sim.py:
```diff
@@ -0,0 +1,508 @@
+    The abstract simulator interface for the motion planner.
+    """
+
+    def __init__(self, sim):
```
types missing

Skylion007 commented on habitat_baselines/motion_planning/mp_sim.py:
```diff
@@ -0,0 +1,508 @@
+        """
+
+
+class HabMpSim(MpSim):
```
types missing

Skylion007 commented on habitat_baselines/motion_planning/mp_sim.py:
```diff
@@ -0,0 +1,508 @@
+        def convert_coll(x):
+            linkA, bodyA = p.getBodyInfo(x[1], physicsClientId=self.pc_id)
+            linkB, bodyB = p.getBodyInfo(x[2], physicsClientId=self.pc_id)
+            bodyA = bodyA.decode("ascii")
```
Hmm, we may want to revisit the API to prevent the need to decode here. It's kinda off for a Python object and easy to fix in PyBind

Skylion007 commented on habitat_baselines/motion_planning/mp_spaces.py:
```diff
@@ -0,0 +1,220 @@
+        self._should_render = should_render
+
+    def convert_state(self, x):
+        return np.array([x[i] for i in range(7)])
```
Type here? If x is an array the list comprehension is unnecesary cn be used as the getter into x.

ASzot replied:
The type is some OMPL internal state vector. I am not sure the proper way to convert it to numpy array.

Skylion007 commented on habitat_baselines/motion_planning/mp_spaces.py:
```diff
@@ -0,0 +1,220 @@
+        pdef.setStartAndGoalStates(js_start, js_end)
+        pdef.setOptimizationObjective(getPathLengthObjWithCostToGo(si))
+
+    def render_start_targ(self, render_dir, subdir, targ_state, suffix="targ"):
```
missing types in the whole file

Skylion007 commented on habitat_baselines/motion_planning/projector/core.py:
```diff
@@ -0,0 +1,300 @@
+import math
+
+import torch
```
Avoid making torch as hard import

Skylion007 commented on habitat_baselines/motion_planning/robot_target.py:
```diff
@@ -0,0 +1,26 @@
+import attr
```
util file is empty?!

Skylion007 reviewed (changes requested):
This API is getting very complicated will definitely need extensive refactoring. Or better integration with Habitat-Sims's pybullet. There are a TON of helper classes here with hard coded conversions/ types/ assumptions of parallel lists etc... 

Marked the most obvious and immediate changes.

Skylion007 commented:
Which file here is the gym wrapper? This is quite a big PR, perhaps the gym wrapper should be split off.

ASzot commented:
> Which file here is the gym wrapper? This is quite a big PR, perhaps the gym wrapper should be split off.

The Gym wrapper is here: https://github.com/facebookresearch/habitat-lab/blob/hab2_motion_planning/habitat_baselines/utils/gym_adapter.py 

I am okay with splitting off the gym wrapper into another PR, but there are anyways a lot of random changes in this PR.

mathfac commented on habitat/tasks/rearrange/actions.py:
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
 
 
+@registry.register_task_action
+class ArmAbsPosAction(SimulatorTaskAction):
```
Doc string needed to say how it's different from other arm actions.

mathfac commented on habitat/tasks/rearrange/actions.py:
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
+class ArmAbsPosAction(SimulatorTaskAction):
+    @property
+    def action_space(self):
+        return spaces.Box(shape=(7,), low=0, high=1, dtype=np.float32)
```
Magical number 7 should have reasonable constant name so we can change it easily when needed and distinguish it from other 7s.

mathfac commented on habitat/tasks/rearrange/actions.py:
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
+
+
+@registry.register_task_action
+class ArmAbsPosKinematicAction(SimulatorTaskAction):
```
Doc string for users unfamiliar with Robotics world.

mathfac commented on habitat/tasks/rearrange/rearrange_sensors.py:
```diff
@@ -70,6 +69,28 @@ def get_observation(self, observations, episode, *args, **kwargs):
 
 
+@registry.register_sensor
+class TargetObjectSensor(MultiObjSensor):
```
```suggestion
class GoalObjectPositionSensor(MultiObjSensor):
```

mathfac commented on habitat/tasks/rearrange/rearrange_sensors.py:
```diff
@@ -70,6 +69,28 @@ def get_observation(self, observations, episode, *args, **kwargs):
+@registry.register_sensor
+class TargetObjectSensor(MultiObjSensor):
+    """
+    This is the ground truth position
```
Example of the good doc string here:
```
Sensor for PointGoal observations which are used in PointGoal Navigation.

    For the agent in simulator the forward direction is along negative-z.
    In polar coordinate format the angle returned is azimuth to the goal.

    Args:
        sim: reference to the simulator for calculating task observations.
        config: config for the PointGoal sensor. Can contain field for
            GOAL_FORMAT which can be used to specify the format in which
            the pointgoal is specified. Current options for goal format are
            cartesian and polar.

            Also contains a DIMENSIONALITY field which specifes the number
            of dimensions ued to specify the goal, must be in [2, 3]

    Attributes:
        _goal_format: format for specifying the goal which can be done
            in cartesian or polar coordinates.
        _dimensionality: number of dimensions used to specify the goal
```

mathfac commented on habitat/tasks/rearrange/rearrange_sensors.py:
```diff
@@ -184,9 +205,9 @@ def _get_uuid(self, *args, **kwargs):
+    def _get_observation_space(self, *args, config, **kwargs):
         return spaces.Box(
-            shape=(9,),
+            shape=(config.JOINT_DIM,),
```
Glad to see constants moving into config.

mathfac commented on habitat/tasks/rearrange/rearrange_sensors.py:
```diff
@@ -329,6 +350,31 @@ def update_metric(self, *args, episode, **kwargs):
 
 
+@registry.register_measure
+class EndEffectorToRestDistance(Measure):
```
Doc string needed describing what resting is about and why it's needed.

mathfac commented on habitat/tasks/rearrange/rearrange_sim.py:
```diff
@@ -402,6 +407,22 @@ def step(self, action):
 
         return obs
 
+    def viz_pos(self, pos, viz_id=None, r=0.05):
```
Longer name and doc string needed.

mathfac reviewed (commented):
First part of the feedback.

mathfac commented:
Let's move GymWrapper to the other PR if possible. That will be easy to review and refer to later.

ASzot commented on configs/tasks/rearrangepick_replica_cad.yaml:
```diff
@@ -37,13 +37,14 @@ TASK:
     JOINT_SENSOR:
         TYPE: "JointSensor"
+        JOINT_DIM: 7
     EE_SENSOR:
```
Resolved some usages, but this will need a follow up PR because "EE" is short for end-effector in many places throughout the code.

ASzot commented:
I tried separating the GymWrapper PR from this PR, however, unfortunately this PR also contains fixes for the task which are necessary for the GymWrapper tests to pass. Separating the PRs will take a lot more time and I don’t think is worth the effort. 

I removed all code related to motion planning from perception in the PyBullet simulator wrapper, this will be added in a much later PR since the code will require major refactoring. I also added a new *Fetch Reach task* which was useful for my debugging. This PR also now fixes issues with the observation space and reward that I only found after training policies with RL.

mathfac commented on habitat/tasks/rearrange/rearrange_task.py:
```diff
@@ -7,25 +7,42 @@
     """
 
+    def overwrite_sim_config(self, sim_config, episode):
+        return merge_sim_episode_with_object_config(sim_config, episode)
```
Should we call parent `overwrite_sim_config` as well here?

mathfac commented on habitat/tasks/rearrange/rearrange_sim.py:
```diff
@@ -402,6 +409,22 @@ def step(self, action):
 
         return obs
 
+    def viz_pos(self, pos, viz_id=None, r=0.05):
```
This looks like util method that should be in Simulator class.

mathfac reviewed (commented):
Next portion.

Skylion007 commented on habitat/tasks/rearrange/actions.py:
```diff
@@ -253,8 +311,9 @@ def apply_ee_constraints(self):
 
-    def set_desired_ee_pos(self, des_rel_pos):
-        self.ee_targ += np.array(des_rel_pos)
+    def set_desired_ee_pos(self, ee_pos):
```
Call signature of this? Iterable? Array?
